### PR TITLE
added the ability to sign in with username

### LINKED ISSUE #17: As a user I need to signin with my username or email
Add the ability to sign in with a username instead of an email address.

### PR DESCRIPTION
**Pull request by aaronmallen**
resolves #17

### CODE REVIEW

File: spec/factories/users.rb [hunk 1]
```diff
@@ -2,11 +2,13 @@
   factory :user do
     confirmed_at Time.now
     name "Test User"
+    sequence(:username, 1) { |n| "testuser#{n}" }
     email "test@example.com"
```
Comment (wasi) [inline spec/factories/users.rb]: email is unique too, so it needs different values as well. I would suggest to use something like https://github.com/sevenwire/forgery for all the values. because you only test for specific values right now.

Comment (aaronmallen) [inline spec/factories/users.rb]: We already have the faker gem installed in the package, I just don't like using it. It's easier to test when I know some of the values. I didn't initially sequence out the email address because on the user model devise is handling everything but the username.  I will sequence out the email address and push a new commit though.  Can you give me a benefit to using faker or forgery vs. just using sequenced hardcoded strings?

**Comment by wasi** [comment]
@aaronmallen "It's easier to test when I know some of the values" could lead you into trouble if the project matures. do not depend on the fact that you know the values and you know where you used them. test should be easy to refactor and to change. if you reference the values in many tests then its hard to refactor. your tests will be tightly coupled to the factory (thats something we always want to avoid. we want to loose coupled classes, methods AND tests)

the difference of sequenced and forgery etc. is not that much. The biggest advantage for me is, that the data has more variety. (compare your email: "testuser#{n}@test.com" with generated emails: john@test.com, frank@mail.co.uk, maggie@email.de -> in your case you do not test for a email, you test for a email ending with .com). same with date, if you hard code the dates, you write a test for a specific date only. Its rare, but I got passing tests because they worked with hardcoded values but failed with other combinations so a bug was undiscovered. I had tests that would fail on the 1. of a month but pass at any other date.

these are very edge cases and a lot comes down to personal preferences. 

the cons for forgery etc. are that they add another dependency. but one I had no problems so far.

**Comment by aaronmallen** [comment]
@wasi that sounds fair,can you tell me the pros to going with forgery over faker (already installed)?

**Comment by aaronmallen** [comment]
@wasi I did some research into it and it would seem the only thing Faker has over forgery is a larger default library. So I'm gonna use faker since it's already installed and ready to go.  I have edited all current factories to use Faker and adjusted specs accordingly.
